staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/compilation.go

```diff
@@ -125,3 +132,39 @@ func Compile(s *schema.Structural, isResourceRoot bool) ([]CompilationResult, er
 func generateUniqueSelfTypeName() string {
 	return fmt.Sprintf("selfType%d", time.Now().Nanosecond())
 }
+
+type celEstimator struct {
+	root *celmodel.DeclType
+}
+
+func (c *celEstimator) EstimateSize(element checker.AstNode) *checker.SizeEstimate {
+	currentNode := c.root
+	// cut off "self" from path, since we always start there
+	for _, name := range element.Path()[1:] {
+		var nextNode *celmodel.DeclType
+		switch name {
+		case "@items", "@values":
+			nextNode = currentNode.ElemType
+			if nextNode == nil {
+				return nil
+			}
+		case "@keys":
+			nextNode = currentNode.KeyType
+			if nextNode == nil {
+				return nil
+			}
+		default:
+			field, ok := currentNode.Fields[name]
+			if !ok {
+				return nil
+			}
+			currentNode = field.Type
+		}
+	}
+	return &checker.SizeEstimate{Min: 0, Max: uint64(currentNode.MaxLength)}
+}
+
+func (c *celEstimator) EstimateCallCost(overloadId string, target *checker.AstNode, args []checker.AstNode) *checker.CostEstimate {
+	// TODO(DangerOnTheRanger): implement in separate PR?
```

Comment:
The TODO can state that this will be implemented to handle the functions introduced in https://github.com/kubernetes/kubernetes/pull/108312 in a future PR.

Reply:
Gotcha, fixed